Update standard K8s version policy (#334)
SovereignCloudStack/standards: Standards/scs-0210-v2-k8s-version-policy.md

version being in development. As soon as a new minor version is officially released,
the oldest version is dropped out of the support period.
Kubernetes supports its releases for around 14 months. 12 of these are the standard
support period. The remaining 2 months are used for:

- CVEs (under the advisement of the Security Response Committee)
- dependency issues (including base image updates)
- critical core component issues

[Kubernetes Support Period](https://kubernetes.io/releases/patch-releases/#support-period)

The Kubernetes releases are planned for about 4 months, which usually results in about
**3 minor** releases per year.
[Kubernetes Release Cycle](https://kubernetes.io/releases/release/#the-release-cycle)

Patches to these releases are provided monthly, with the exception of the first patch,
which is usually provided 1-2 weeks after the initial release.
[Patch Release Cadence](https://kubernetes.io/releases/patch-releases/#cadence)

## Motivation

Kubernetes is a living, fast-paced project, which follows a pre-defined release cycle.
This enables forward planning with regards to releases and patches, but also implies a
necessity to upgrade to newer versions quickly.

> We want to achieve an up to date policy, which means that providers should be mostly in sync with the upstream and don't

Reviewer: Language hint:
up-to-date (with dashes).

> fall behind the official Kubernetes releases. In our opinion, this should be achievable, since new versions are released

Reviewer: "In our opinion" is not the language I would expect in a standard.
Suggestion:
This is achievable since new versions are released ...
(without comma)

Author: Fixed.

periodical on a well communicated schedule. This should enable processes to be set up around this.

> Being up to date ensures, that security issues and bugs are addressed and new features are made available when using SCS compliant clusters.

Reviewer: no comma in front of that

Author: Fixed.

On the other hand, providers should support minor Kubernetes versions at least as long as

> the official support period is, since users could depend on specific versions.

Reviewer: OTOH, since users may depend on specific versions, providers should support minor Kubernetes versions *for the duration of the official support period*.

Author: Agree. I wrote this into the document in the latest commit.

This could be a bit contradicting in some cases, because the necessity to upgrade versions

> would theoretically drop the support period after that upgrade period.

Reviewer: I don't get it. What i contradicting?
Month 0: New minor k8s version N
Month ~2: Introduction of the version by service provider
Month 12: End of generic upstream maintenance.
Month 12: Recommendation by providers to users to upgrade (to version N+3 (if already provided), N+2 or N+1)
Month 14: End of upstream security fixes
Month 14: Final warning and end of support for minor by provider

Where's the contradiction?

Author: See my comment https://github.com/SovereignCloudStack/standards/pull/334#issuecomment-1691253644

Author: Changed this section, so closing the conversation.

It is therefore more reasonable, to recommend a version upgrade and make the version
support period a necessity.

## Decision

In order to keep up-to date with the latest Kubernetes features, bug fixes and security improvements,
the provided Kubernetes versions should be kept up to date with the upstream.

> - It is RECOMMENDED, that the latest minor version is provided no later than 4 months after release.

Reviewer: We wanted to have stronger language here in the past.
Providers MUST provide the latest minor version no later than ...

With that language, we can make this a hard SCS-compatible Container certification requirement.

Author: Changed this section, so closing the conversation.

- It is RECOMMENDED, that the latest patch version is provided no later than 1 week after release.

> - It is RECOMMENDED, that this time period should be even shorter for patches that target critical CVEs (CVSS >= 8). It is probably recommended to have a patch period around 1 to 3 days for these cases.

Reviewer: You could leave out the last sentence. If you want to keep it, remove the "probably"

Author: I agree, but that was also more of an example/proposal, because I don't know what would be the best timescale for such a critical patch.

Author: Changed this section, so closing the conversation.

> - It is RECOMMENDED, that new versions SHOULD be tested before being rolled out on an infrastructure. How this testing will be done won't be defined in this document.

Reviewer: We could add that at least the CNCF conformance testsuite must be passed.

Again on RECOMMENDED vs MUST: The detail with faster patching for CVEs probably should stay in the RECOMMENDED state, but I see no reason not to mandate the rest.

Author: Added the MUST keyword on the critical patch sentence.

I would leave the last sentence in and the document on `Draft` until tests are written and we can write in what tests are used.

At the same time, provider should support Kubernetes versions at least as long as the official sources as mentioned
in the Kubernetes Support Period document.

> - Kubernetes versions SHOULD be supported as long as the official sources support them. At the moment, this support period

Reviewer: Isn't this statement a duplication of the previous paragraph?

Author: Depends on which paragraph you mean here. I thought the "decision" paragraph should contain the actual standard decisions, which IMO could "duplicate" something from a previous paragraph.

Author: Changed this section, so closing the conversation.

---
title: SCS K8S Version Policy
type: Standard
status: Draft
stabilized_at: 2023-08-16
track: KaaS
---

## Introduction

The Kubernetes project maintains multiple release versions including their patched versions.
In the project, the three most recent minor releases are actively maintained, with a fourth
version being in development. As soon as a new minor version is officially released,
the oldest version is dropped out of the support period.
Kubernetes supports its releases for around 14 months. 12 of these are the standard
support period. The remaining 2 months are end-of-life support for things like:

> - CVEs (under the advisement of the Security Response Committee)

Reviewer 2: What does "CVE" stand for?

Reviewer: https://en.wikipedia.org/wiki/Common_Vulnerabilities_and_Exposures

Author: Thank you for providing the link, I will put in a glossary section next week.

- dependency issues (including base image updates)
- critical core component issues

> More information can be found under [Kubernetes Support Period].

Reviewer 2: There is a hyperlink missing, isn't it?

Reviewer: https://kubernetes.io/releases/
https://endoflife.date/kubernetes

Author: The hyperlinks are at the end of the document in their own section.

Author: Check [here](https://github.com/SovereignCloudStack/standards/pull/334#discussion_r1341266362)

The Kubernetes releases are planned for about 4 months, which usually results in about

> **3 minor** releases per year [Kubernetes Release Cycle].

Reviewer 2: d
```suggestion
**3 minor** releases per year [Kubernetes Release Cycle](https://kubernetes.io/releases/release/).
```

Reviewer 2: I just recognized, that links are listed at the end of the document. Is this a common practice in SCS standards?

Reviewer: No. Linking docs that explain a statement or term is best be done in place. A link I would click if I need an explanation or definition or want to see the source of truth for a statement.

A references section at the end is more for further reading or background reading (in case I'm new to the topic).
Also link to related content, e.g. related standards.

Author: Should we really put links in different locations depending on their relation to the standard? I would rather try to put everything in the final section and check if the link nevertheless works in the document itself.

Reviewer 2: In favor of readability, I would add links in place. And put, as @garloff mentioned, additional documents, not related to a special term, at the bottom of the document.

Author: Put the links in the text.

necessity to upgrade to newer versions quickly, since these often include new features,
important security updates or especially if a previous version falls out of the support
period window.

We want to achieve an up-to-date policy, meaning that providers should be mostly in
sync with the upstream and don't fall behind the official Kubernetes releases.
This is achievable, since new versions are released periodical on a well communicated
schedule, enabling providers and users to set up processes around it.
Being up to date ensures that security issues and bugs are addressed and new features
are made available when using SCS compliant clusters.

It is nevertheless important to at least support all Kubernetes versions that are still
inside the support period, since users could depend on specific versions or may need
longer to upgrade their workloads to a newer version.

The standard therefore should provide a version recency policy as well as a support
window period.

## Decision

In order to keep up-to date with the latest Kubernetes features, bug fixes and security improvements,
the provided Kubernetes versions should be kept up to date with the upstream.

> - The latest minor version SHOULD be provided no later than 4 months after release.

Reviewer: Why is this not a MUST?
4 months is plenty of time to get the latest minor release built, validated, fixed (if needed, which is normally not the case) and published.

Author: To answer all of the MUST questions, I had a grave misunderstanding with the way how MUST and SHOULD are used. In my mind, they were equal, but after reading the RFC document regarding MUST / SHOULD / ... I can see my mistake.
And you're right, all of the ones you pointed out should be MUST.

> - The latest patch version SHOULD be provided no later than 1 week after release.

Reviewer: I believe we can make this a MUST as well.

Author: See https://github.com/SovereignCloudStack/standards/pull/334#discussion_r1392522638

- The latest patch version SHOULD be provided no later than 1 week after release.

> - This time period MUST be even shorter for patches that target critical CVEs (CVSS >= 8).

Reviewer 2: How short MUST support time for critical patches be?

Reviewer: The statement here says MUST be shorter than one week (i.e. < 168 hours), and the next sentence goes on to recommend smaller than 2d (<48hrs).

Author: As @garloff said, the next sentence provides the information about the support time for the critical patches.

>   It is RECOMMENDED to provide the version in a 2 day time period after release.

Reviewer 2: A new "version" of what: patch or minor release?

Reviewer: ```suggestion
  It is RECOMMENDED to provide a new patch version in a 2 day time period after release in this case.
```

Author: Updated the line, but changed the second suggestion a bit since it sounded not quite right.

> - New versions SHOULD be tested before being rolled out on productive infrastructure.

Reviewer 2: A new "version" of what: patch or minor release?

Reviewer: Both.
We could add here that they need to pass the CNCF E2E tests at least.

Author: Added this line.

> - New versions SHOULD be tested before being rolled out on productive infrastructure;

Reviewer: MUST?

Author: See https://github.com/SovereignCloudStack/standards/pull/334#discussion_r1392522638

>   at least the CNCF E2E tests should be passed beforehand.

Reviewer: could be a MUST as well (and if we want to be generous, we could allow failed e2e tests to be listed in a release note rather than requiring 100% pass rate).

Author: See https://github.com/SovereignCloudStack/standards/pull/334#discussion_r1392522638

Also I think you're right about the generosity of passed tests, but I'm not really sure how to properly display this in Release notes.

> At the same time, provider should support Kubernetes versions at least as long as the

Reviewer: Again, I would make this a MUST.

in the Kubernetes Support Period document.

- Kubernetes versions SHOULD be supported as long as the official sources support them.
  At the moment, this support period is 14 months.

> - It is RECOMMENDED to not support versions after this period in order to not encourage
>   usage of out-of-date versions.

Reviewer 2: As line 57 allows to delay latest minor version for 4 month, I would take this into account with respect to shut of oldest minor version. I can image, that customers may not adapt their applications to latest minor version, but may skip ones. Hence, I would rephrase this sentence in order to not remove older verisons, before new versions will be supported.

```suggestion
- It is RECOMMENDED to not support versions after this period in order to not encourage
  usage of out-of-date versions, but not before the latest minor version is supported.
```

Reviewer: If I do the math:
1.N is released and provided in month 0, extended support till month 14
1.N+1 is released in month 4 and provided in month 8, extended support till month 18
1.N+2 is released in month 8 and provided in month 12, extended support till month 22
So, skipping one level from 1.N to 1.N+2 is still possible with a 2months overlap window.
Sidenote: Technically, a rolling upgrade will need to do the intermediate step (but of course that could be very short-lived).
I would not complicate the policy, as skipping one version does not require it.

Author: I would also not add this clause, since it complicates things as @garloff said, and the versions are provided in regular time intervals. Also we can't control what customers want to do with their applications; as long as the providers give them access to the latest versions (as the standard says), the can decide themselves what to do.

> - Kubernetes versions SHOULD be supported as long as the official sources support them.

Reviewer: Hmm, looks duplicate to the sentence before.
And should be MUST from my PoV.
